cuda/solver/common_trs_kernels.cu

```diff
@@ -0,0 +1,1910 @@
+        } else {
+            if (is_upper) {
+                sptrsmebrcnm_kernel<true><<<grid_size, block_size>>>(
+                    matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                    as_cuda_type(matrix->get_const_values()),
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    nan_produced.get_data(), atomic_counter.get_data(), m,
+                    unit_diag);
+            } else {
+                sptrsmebrcnm_kernel<false><<<grid_size, block_size>>>(
+                    matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                    as_cuda_type(matrix->get_const_values()),
+                    as_cuda_type(b->get_const_values()), b->get_stride(),
+                    as_cuda_type(x->get_values()), x->get_stride(), n, nrhs,
+                    nan_produced.get_data(), atomic_counter.get_data(), m,
+                    unit_diag);
+            }
+        }
+#if GKO_VERBOSE_LEVEL >= 1
+        if (exec->copy_val_to_host(nan_produced.get_const_data())) {
+            std::cerr << "Error: triangular solve produced NaN, either not all "
+                         "diagonal "
+                         "elements are nonzero, or the system is very "
+                         "ill-conditioned. "
+                         "The NaN will be replaced with a zero.\n";
+        }
+#endif  // GKO_VERBOSE_LEVEL >= 1
+    }
+};
+
+
+template <typename ValueType, typename IndexType, bool is_upper>
+__global__ void sptrsvelcr_kernel(
+    const IndexType* const rowptrs, const IndexType* const colidxs,
+    const ValueType* const vals, const ValueType* const b,
+    const size_type b_stride, ValueType* const x, const size_type x_stride,
+    const IndexType* const levels, const IndexType sweep, const IndexType n,
+    const IndexType nrhs, bool unit_diag)
+{
+    const auto gid = thread::get_thread_id_flat<IndexType>();
+    const auto row = gid / nrhs;
+    const auto rhs = gid % nrhs;
+
+    if (row >= n) {
+        return;
+    }
+
+    if (levels[row] != sweep) {
+        return;
+    }
+
+    const auto row_start = is_upper ? rowptrs[row + 1] - 1 : rowptrs[row];
+    const auto row_end = is_upper ? rowptrs[row] : rowptrs[row + 1] - 1;
+    const auto row_step = is_upper ? -1 : 1;
+
+    auto sum = zero<ValueType>();
+    IndexType i = row_start;
+    for (; i != row_end; i += row_step) {
+        const auto dependency = colidxs[i];
+        if (is_upper ? dependency <= row : dependency >= row) {
+            break;
+        }
+
+        sum += x[dependency * x_stride + rhs] * vals[i];
+    }
+
+    const auto diag = unit_diag ? one<ValueType>() : vals[i];
+    const auto r = (b[row * b_stride + rhs] - sum) / diag;
+    x[row * x_stride + rhs] = r;
+}
+
+
+template <typename IndexType, bool is_upper>
+__global__ void level_generation_kernel(const IndexType* const rowptrs,
```

Comment:
maybe we can find a way to merge this kernel with the numerical one (same for legacy) with the right generalizations of + and *?